### PR TITLE
Serviceutils dataclasses refactoring

### PR DESCRIPTION
**Pull request by MatthewBaggins**
I changed classes in serviceutils.py into dataclasses with slots. I think this improves readability a bit and explicit information about types of attributes makes it possible to use static type checkers (e.g. pylint, mypy) to prevent runtime errors.

### CODE REVIEW

File: utilities/serviceutils.py
Comment (Aprillion) [inline utilities/serviceutils.py:120]: hmm, static analysis tools shouldn't need the [forward reference](https://peps.python.org/pep-0484/#forward-references) type hint syntax with quotes, do we use any tool that needs it?

if yes, have you checked if other places in the code need it too, please?

**Review by Aprillion — COMMENTED** [review]
I am used to Java Lombok and Spring annotations that add a lot of magic to reduce OOP boilerplate code, so `@dataclass` and `field()` are readable to me, but it is one more layer of abstraction to imagine the autogenerated __init__ from the decorator and I'm not entirely sure this feels "pythonic" to me 🤔 

I think TJ, Chris, and other full time programmers might be fine with this change too, less sure if we get volunteer contributors with other backgrounds

@robertskmiles - do we want to embrace this brave new Python OOP world, is this syntax comfortable enough for you to update stuff in the future? if you see `@dataclass`, is it easy to search for and learn what it means?

if this was already discussed during office hours as a good approach, let me know, I will be happy to approve ;)

File: utilities/serviceutils.py
Comment (MatthewBaggins) [reply utilities/serviceutils.py:120]: If an attribute of a class has the same type as its class it needs to be quoted, otherwise it's going to throw an error. Same if you want to type annotate an argument or return type of one of its methods (e.g. `def __eq__(self, other: "C")`. You can omit this need by just adding `from __future__ import annotations` at the top of the file and I'm with that option too.

Comment (Aprillion) [reply utilities/serviceutils.py:120]: ah, `@dataclass` itself will throw an error or which tool? type systems inside VS Code and IntelliJ both seemed fine without the quotes AFAICT 😅

Comment (MatthewBaggins) [reply utilities/serviceutils.py:120]: > it is one more layer of abstraction to imagine the autogenerated init from the decorator

I think defining what attributes a class should have and having `__init__` autogenerated from that increases code clarity. I don't think imagining what arguments should be passed to `__init__` is a problem. You just make sure to add all `field`s with `init=False` below those with (implicit) `init=True` and whenever you create an instance of this class or call `super()__init__`, your IDE displays it to you.

Comment (MatthewBaggins) [reply utilities/serviceutils.py:120]: > ah, `@dataclass` itself will throw an error or which tool? type systems inside VS Code and IntelliJ both seemed fine without the quotes AFAICT sweat_smile

`@dataclass` itself

Screenshot from 2022-09-17 11-04-50

**Comment by tayler6000** [comment]
> I think TJ, Chris, and other full time programmers might be fine with this change too, less sure if we get volunteer contributors with other backgrounds

Personally, I don't like the look of it much, and I don't feel it's very pythonic either, but with that said, I'm willing to adapt to the needs and wants of the team as a whole.

**Comment by jakkdl** [comment]
Chiming in as experienced python dev, `@dataclass` is definitely pythonic imo and I think it looks cleaner and is easy to adapt to even if you've not seen it before (meanwhile `__init__` is kinda magical). Big fan of type hints as well.

**Comment by tayler6000** [comment]
> Chiming in as experienced python dev, `@dataclass` is definitely pythonic imo and I think it looks cleaner and is easy to adapt to even if you've not seen it before (meanwhile `__init__` is kinda magical). Big fan of type hints as well.

I'm also a big fan of type hints and unit tests, which this project lacks lots off, slowly but surely I've been sneaking them in.  I don't know anything when it comes to dataclasses though.  If you can confirm that this code (before and after refactor) is equivalent feel free to approve it @jakkdl but we should hold off on the merge until we get a response from @robertskmiles and @Aprillion.

File: utilities/serviceutils.py
Comment (tayler6000) [reply utilities/serviceutils.py:120]: Oh how I long for `from __future__ import annotations` to just be the default.  Alternatively to quoting we actually could just import this.  Are there any drawbacks to doing so?

File: utilities/serviceutils.py
Comment (jakkdl) [inline utilities/serviceutils.py:16]: why use `field` with `init=False` when changing not to use `hasattr`?
I'd just default `_role` to `None`, though it looks like other parts of the code uses `hasattr` so removing all of those is maybe a separate PR - in which case you can simplify this PR by not changing the logic.

**Comment by Aprillion** [comment]
I'm fine with this change as individual contributor, but it's a big code-style change IMHO, so not immediately approving without confirmation this was discussed with more [contributors](https://github.com/StampyAI/stampy/graphs/contributors) .. I missed a lot of meetings, so genuinely don't know whether this topic was open before this PR or this is the first time we are thinking about it and just need more time to collectively decide 😂

**Comment by robertskmiles** [comment]
I basically haven't had any involvement with the Service Modules code, so I don't really have an opinion, I'm happy to go with whatever. One thing I'll say though is it's good to have the codebase be consistent, so if this is more generally applicable throughout the codebase I think I'd want it to be done everywhere or nowhere

**Comment by jakkdl** [comment]
@robertskmiles Agreed. Looking at the classes that were changed to `@dataclass` that looks decently fitting, though when several dunder classes that would otherwise be generated has to be implemented the case is weaker.

If one wants a bigger rewrite where this syntax is consistently used, without the implication that classes are just data containers, there's https://github.com/python-attrs/attrs (among others) which is decently widely used and was an inspiration for dataclasses.

File: utilities/serviceutils.py
Comment (tayler6000) [reply utilities/serviceutils.py:16]: Ideally you'd update the other hasattr's in this PR.

**Review by Aprillion — APPROVED** [review]
Matthew won't have much time next few weeks (according to office hours call) => IMHO we can merge as-is and ask for improvements in a new issue, perhaps.

**Pull request merged**
